Add the `outline` and `check-crate-conf` commands

norosa opened the pull request:
I am currently working on <https://github.com/crate/crate/pull/11277>, and after I converted `crate.yml` to use RST-style heading syntax, I immediately noticed a problem. The outline of the file looked like this:

    Basic setup
      Memory
      Persistence
      Networking
      Clustering
    Advanced settings
      JMX monitoring plugin
      Authentication
        Host-based access
        Trust-based authentication
      User-defined functions (UDFs)
      Secured communications (SSL/TLS)
        Optional configuration for the truststore
      Cluster settings
        Graceful stop
        Routing allocation
      Node settings
      Paths
      Memory
      Network and HTTP
      Metadata gateway
      Recovery throttling
      Discovery settings
      Sharding
      Field-data circuit breaker
      Threadpools
      Metadata
      Garbage collection (GC) logging
      SQL
      Query circuit breaker
      Usage-data collector (UDC)
      Backup and restore (aka snapshots)
      PostgreSQL wire transport

The "Basic setup" section is okay because it does not attempt to mirror anything in the documentation. However, the "Advanced settings" section is supposed to list all CrateDB configuration options (it does not list all options, but that is potentially another topic for another day).

The most immediate problem with this outline is that the CrateDB documentation splits the configuration options into two major sections:

- Node-specific settings
  <https://crate.io/docs/crate/reference/en/latest/config/node.html>

- Cluster-wide settings
  <https://crate.io/docs/crate/reference/en/latest/config/cluster.html>

In the `crate.yml` file, we have:

- A "Cluster settings" section with two subsections: "Graceful stop" and "Routing allocation"

- A "Node settings" section with no subsections

These two sections appear in reverse order and are positioned halfway down the file. The rest of the sections resemble a seemingly random assortment of subsections from both the "Node-specific settings" and "Cluster-wide settings" documents.

The result is a confusing and misleading mess of information. However, the cause of this problem is straightforward enough: The two "sources of truth" (re what the config options are) have had over half a decade to drift apart.

For comparison, below are the outlines of the two sections in the docs.

Node-specific settings:

    Basics
    Node types
    Read-only node
    Hosts
    Ports
    Paths
    Plug-ins
    CPU
    Memory
    Garbage collection
    Authentication
      Trust authentication
      Host-based authentication
          HBA entries
    Secured communications (SSL/TLS)
    Cross-origin resource sharing (CORS)
    Blobs
    Repositories
    Queries
    JavaScript language
    Custom attributes

Cluster-wide settings:

    Non-runtime cluster-wide settings
    Collecting stats
    Shard limits
    Usage-Data Collector
    Graceful stop
    Bulk operations
    Discovery
      Unicast host discovery
        Discovery via DNS
        Discovery on Amazon EC2
        Discovery on Microsoft Azure
    Routing allocation
      Awareness
      Balanced shards
      Cluster-wide allocation filtering
      Disk-based shard allocation
    Recovery
    Memory management
    Circuit breakers
      Query circuit breaker
      Field-data circuit breaker
      Request circuit breaker
      Accounting circuit breaker
      Stats circuit breakers
      Total circuit breaker
    Thread pools
      Settings for fixed thread pools
    Metadata
      Metadata gateway

As I started to attempt to bring `crate.yml` in line with the documentation, I realized it would be difficult to do this by hand. I also realized that the two sources of truth would quickly drift apart again. An unfortunate consequence of this practically unavoidable violation of the principle of Do Not Repeat Yourself.

I have used `diff` to verify the contents of tarballs before, and I realized that I could use the same method to compare the two sources of truth if only I could generate outlines of the headings.

I wrote the `outline` script added by this commit to satisfy that wish. Read the docstring in the file for details (there is an example showing how the program can be used with `diff` to satisfy my use-case).

By way of demonstration, I generated the outlines in this commit message (above) using the `outline` program.

My original plan was to commit this file to the primary CrateDB repository, with the thought that we might eventually use it for automated testing. However, I realized that a secondary use for this program is as a validation tool for our RST style guide more generally.

For example, when I first ran `outline` on `cluster.rst` in the docs, I noticed that some sections were at the wrong level (e.g., one level too deep). The standard RST parser can only tell you if your use of underlines is inconsistent within one document. It cannot tell you if your use of underlines is inconsistent between documents.

norosa commented:
I have added a second commit to this PR. I wrote this script to help me put the `outline` program to use while working on `crate.yml`. it occurred to me that this script may be helpful in the future, so I shaped it up a bit and added some documentation so other people can figure out a) what it is for, and b) how to use it

here's what it looks like in action:

![image](https://user-images.githubusercontent.com/23469/115575765-28194800-a2c3-11eb-88f7-c405005fa914.png)

this screenshot shows the current difference between the "cluster settings" section of the `crate.yml` file and the "cluster settings" page in the CrateDB documentation. ideally, these two "sources of truth" align

amotl commented on helpers/check-crate-conf:
```diff
@@ -0,0 +1,113 @@
+fi
+
+# Generate two outlines in the previously created temporary directory
+outline "${rst_file}" > "${temp_dir}/${rst_outline}"
+outline "${crate_yml}" "${yml_heading}" "#" > "${temp_dir}/${yml_outline}"
```
```suggestion
rst-outline "${rst_file}" > "${temp_dir}/${rst_outline}"
rst-outline "${crate_yml}" "${yml_heading}" "#" > "${temp_dir}/${yml_outline}"
```

amotl commented on helpers/outline:
```diff
@@ -0,0 +1,167 @@
+#!/usr/bin/python3
+
+r"""
+Outline a document's heading structure.
```
```suggestion
Outline the heading structure of a reStructuredText document.
```

amotl commented on helpers/outline:
```diff
@@ -0,0 +1,167 @@
+
+Synopsis:
+
+    $ ./outline FILENAME [ ROOT_HEADING ] [ LEFT_STRIP ]
```
```suggestion
    $ ./rst-outline FILENAME [ ROOT_HEADING ] [ LEFT_STRIP ]
```

amotl commented on helpers/outline:
```diff
@@ -0,0 +1,167 @@
+
+  First, generate an outline of both files:
+
+      $ ./headings.py path/to/crate/docs/config/node.rst \
```
```suggestion
      $ ./rst-outline path/to/crate/docs/config/node.rst \
```

amotl commented on helpers/outline:
```diff
@@ -0,0 +1,167 @@
+      $ ./headings.py path/to/crate/docs/config/node.rst \
+            > outline-rst.txt
+
+      $ ./headings.py path/to/crate/app/src/main/dist/config/crate.yml \
```
```suggestion
      $ ./rst-outline path/to/crate/app/src/main/dist/config/crate.yml \
```

amotl reviewed (changes requested):
Hi Naomi,

nice little tooling you built here, thanks for sharing. Besides the inline remarks (let's also call this program `rst-outline`?), I would like to see the Python program layout adhere to minimal coding standards like wrapping all code into a section like `if __name__ == "__main__":`.

While that might sound overengineered for little programs like this utility, it is really important for any scenarios like test discovery tools scanning a whole filesystem tree and accidentally executing Python code just by trying to evaluate a Python file.

I suggest to use the following program layout for your specific use case:

```python
def main(target_filename, root_heading=None, left_strip=None):
    # Your whole program modulo the code seen below goes here.
    pass


def exit_help():
    """Print help message and exit."""
    print(__doc__.strip())
    sys.exit(0)


if __name__ == "__main__":

    try:
        target_filename = sys.argv[1]
    except IndexError:
        exit_help()

    try:
        root_heading = sys.argv[2]
        root_re = re.compile("^ *{} *$".format(re.escape(root_heading)))
    except IndexError:
        root_heading = None

    try:
        left_strip = sys.argv[3]
        left_strip_re = re.compile("^ *{} *".format(re.escape(left_strip)))
    except IndexError:
        left_strip = None

    main(target_filename, root_heading, left_strip)
```

With kind regards,
Andreas.

amotl commented on helpers/outline:
```diff
@@ -0,0 +1,168 @@
+
+    $ ./outline FILENAME [ ROOT_HEADING ] [ LEFT_STRIP ]
+
+Scan a file (`FILENAME`) for headings marked up using the heading structure
```
```suggestion
Scan a file (``FILENAME``) for headings marked up using the heading structure
```

amotl commented on helpers/outline:
```diff
@@ -0,0 +1,168 @@
+see <https://github.com/crate/crate-docs/blob/main/style/rst.rst>.
+
+NOTE: The RST markup style is sometimes used in non-RST files (e.g.,
+`crate.yml`). This script can be used on those files too.
```
```suggestion
``crate.yml``). This script can be used on those files too.
```

amotl commented on helpers/outline:
```diff
@@ -0,0 +1,168 @@
+the remaining headings structure (printed to STDOUT) with heading levels
+indicated by horizontal indentation.
+
+If you specify a value for `ROOT_HEADING`, the first subheading that matches
```
```suggestion
If you specify a value for ``ROOT_HEADING``, the first subheading that matches
```

amotl commented on helpers/outline:
```diff
@@ -0,0 +1,168 @@
+that value will be taken as the root heading and the program will only
+outline the direct descendants of that subheading.
+
+If you specify a value for `LEFT_STRIP`, the program will strip that value from
```
```suggestion
If you specify a value for ``LEFT_STRIP``, the program will strip that value from
```

amotl commented on helpers/outline:
```diff
@@ -0,0 +1,168 @@
+
+If you specify a value for `LEFT_STRIP`, the program will strip that value from
+the left-hand side of every line. You can use this argument to process headings
+that are commented out (e.g., set `LEFT_STRIP` to `#`).
```
```suggestion
that are commented out (e.g., set ``LEFT_STRIP`` to `#`).
```

norosa commented:
@amotl thanks for the review!! I made the following changes based on your suggestions:

- renamed the script `outline-headings`. I didn't want to call it `outline-rst` because that implies that the script is only intended for RST files. I think the documentation in the script is sufficient with regards to RST being the base style expected for heading markup

- converted the script to modularize the functionality as suggested

- I didn't change the quoting of literal values in the docstrings. the double backtick style is an RST specific thing. for normal CLI scripts I prefer to use the standard single backtick style used by other programs for UNIX-like systems

amotl reviewed (approved):
> Converted the script to modularize the functionality as suggested.

Excellent, thank you!
